File: Iteration-2/src/Student.java
import java.util.ArrayList;

/**
 * Student class includes some fields and methods related to the
 * student in our domain.
 *
 * @author canba
 *
 */
public class Student {

    private String name;
    private StudentID studentID;
    private int registrationOrder;
    private ArrayList<PreviousCourseGrade> grades;
    private ArrayList<Course> takenCourses;
    private int year;
    private Advisor advisor;
    private WeeklySchedule schedule;
    private Transcript transcript;
    private RegistrationController registrationController;
    private String logString = "";

    public Student(String name, int year, int registrationOrder, RegistrationController registrationController) {
        this.name = name;
        this.year = year;
        this.registrationOrder = registrationOrder;
        this.grades = new ArrayList<>();
        this.takenCourses = new ArrayList<>();
        this.studentID = new StudentID(this);
        this.registrationController = registrationController;
        this.transcript = new Transcript(this);
        this.schedule = new WeeklySchedule(this);
    }

    /**
     * This method loads previous course list
     *
     * @return This returns an arraylist of previous courses of the student.
     *
     */
    private ArrayList<Course> getPreviousCourses() { //
        ArrayList<Course> previousCourses = new ArrayList<>();
        for (PreviousCourseGrade g : grades) {
            previousCourses.add(g.getCourse());
        }
        return previousCourses;
    }

    public boolean isPassed(Course course) { // This method checks whether student passed to course or not
        ArrayList<Course> passedCourses = getPreviousCourses();

        for (Course c : passedCourses) {
            if (course.getCourseCode().equals(c.getCourseCode())) { // If the previousCourses list contains the course's code that means student passed the course and method returns true
                return true;
            }
        }
        return false;
    }

    public void newCourses(CourseSection courseSection) {
        schedule.addToSchedule(courseSection);
        takenCourses.add(courseSection.getCourse());
    }

    public void getGrade(Course course) { // This method randomly appoints a grade to a passed courses (Grade should be more than 50 according to the departments rules)
        int grade = (int) (Math.random() * 51) + 50;
        grades.add(new PreviousCourseGrade(course, grade));
        logString += "\n" + course.getCourseCode() + ": " + grades.get(grades.size() - 1).convertLetterGrade();
    }

    public void selectAndSendToAdvisor() {
        ArrayList<CourseSection> newCourseList = registrationController.createNewCourseList(this);
        setLogString("\n\nCourse List: \n");
        for (CourseSection c : newCourseList) {
            setLogString(c.getCourseSectionCode() + " ");
        }
        setLogString("\n");
        for (CourseSection c: newCourseList) {
            advisor.approveCourse(this, c);;
        }

        setLogString("\n\nTaken Courses: \n");
        for (Course c: takenCourses) {
            setLogString(c.getCourseCode() + " ");
        }
    }

    public String getLogString() {
        return logString;
    }

    public void setLogString(String logString) {
        this.logString += logString;
    }

    public String getName() {
        return name;
    }

    public StudentID getStudentID() {
        return studentID;
    }

    public ArrayList<PreviousCourseGrade> getGrades() {
        return grades;
    }

    public ArrayList<Course> getTakenCourses() {
        return takenCourses;
    }

    public int getYear() {
        return year;
    }

    public Advisor getAdvisor() {
        return advisor;
    }

    public void setAdvisor(Advisor advisor) {
        this.advisor = advisor;
    }

    public WeeklySchedule getSchedule() {
        return schedule;
    }

    public void setSchedule(WeeklySchedule schedule) {
        this.schedule = schedule;
    }

    public Transcript getTranscript() {
        return transcript;
    }
    public int getRegistrationOrder() {
        return registrationOrder;
    }

    public void setRegistrationOrder(int registrationOrder) {
        this.registrationOrder = registrationOrder;
    }

    public void setYear(int year) {
        this.year = year;
    }

}

File: Python Iteration/src/Student.py
from random import randrange

import Advisor
import Transcript
import WeeklySchedule
from PreviousCoursesGrade import PreviousCoursesGrade
from StudentID import StudentID


class Student(object):
    def __init__(self, name, year, registration_order, reg_controller):
        self.__name = name
        self.__year = year
        self.__reg_order = registration_order
        self.__reg_controller = reg_controller
        self.__grades = []
        self.__studentID = StudentID(self)
        self.__schedule = WeeklySchedule.WeeklySchedule(self)
        self.__transcript = Transcript.Transcript(self)
        self.__log = ""
        self.__taken_courses = []
        self.__advisor = Advisor.Advisor

    def get_previous_courses(self):
        previous_courses = []
        for prev_course_grade in self.__grades:
            previous_courses.append(prev_course_grade.course)
        return previous_courses

    def is_passed(self, course):
        for c in self.get_previous_courses():
            if course.course_code == c.course_code:
                return True
        return False

    def new_courses(self, course_section):
        self.__schedule.add_to_schedule(course_section)
        self.__taken_courses.append(course_section.course)

    def get_grade(self, course):
        prev_course_grade = randrange(51) + 50
        self.__grades.append(PreviousCoursesGrade(course, prev_course_grade))
        self.set_log("\n" + course.course_code + ": " + self.__grades[len(self.__grades) - 1].convert_letter_grade())

    def select_and_send_to_advisor(self):
        new_course_list = self.__reg_controller.create_new_course_list(self)
        self.set_log("\n\nCourse List: \n")
        for course_section in new_course_list:
            self.set_log(course_section.course_section_code + " ")
            self.log = "\n"
        for course_section in new_course_list:
            self.__advisor.approve_course(self, course_section)
        self.set_log("\n\nTaken Courses: \n")
        for course in self.__taken_courses:
            self.set_log(course.course_code + " ")

    def set_log(self, log_message):
        self.__log += log_message

    @property
    def advisor(self):
        return self.__advisor

    @advisor.setter
    def advisor(self, advisor):
        self.__advisor = advisor

    @property
    def year(self):
        return self.__year

    @property
    def transcript(self):
        return self.__transcript

    @property
    def student_id(self):
        return self.__studentID

    @property
    def name(self):
        return self.__name

    @property
    def grades(self):
        return self.__grades

    @grades.setter
    def grades(self, grade):
        self.__grades.append(grade)

    @property
    def log(self):
        return self.__log

    @log.setter
    def log(self, log):
        self.__log += log

    @property
    def schedule(self):
        return self.__schedule

    @property
    def reg_order(self):
        return self.__reg_order
